pkg/controllers/status/cluster_status_controller.go

```diff
@@ -131,20 +130,22 @@ func (c *ClusterStatusController) syncClusterStatus(cluster *clusterv1alpha1.Clu
 		return c.setStatusCollectionFailedCondition(cluster, currentClusterStatus, fmt.Sprintf("failed to create a ClusterClient: %v", err))
 	}
 
-	var online, healthy bool
-	// in case of cluster offline, retry a few times to avoid network unstable problems.
-	// Note: retry timeout should not be too long, otherwise will block other cluster reconcile.
-	err = wait.PollImmediate(clusterStatusRetryInterval, clusterStatusRetryTimeout, func() (done bool, err error) {
-		online, healthy = getClusterHealthStatus(clusterClient)
-		if !online {
-			klog.V(2).Infof("Cluster(%s) is offline.", cluster.Name)
-			return false, nil
-		}
-		return true, nil
-	})
-	// error indicates that retry timeout, update cluster status immediately and return.
-	if err != nil {
-		klog.V(2).Infof("Cluster(%s) still offline after retry, ensuring offline is set.", cluster.Name)
+	online, healthy := getClusterHealthStatus(clusterClient)
+	c.Lock()
+	if online && healthy {
+		c.ClusterFailureCounter[cluster.Name] = 0
+	} else if c.ClusterFailureCounter[cluster.Name] < c.ClusterFailureThreshold {
+		// retry a few times to avoid network unstable problems.
+		c.ClusterFailureCounter[cluster.Name]++
+		c.Unlock()
+		return controllerruntime.Result{RequeueAfter: c.ClusterStatusUpdateFrequency.Duration}, nil
+	}
+	c.Unlock()
```

Comment:
Much better than the last commit.
Can we finish it by this PR? Please consider adding a unit test for it.